src/main/java/com/powsybl/openloadflow/ac/AcloadFlowEngine.java

```diff
@@ -128,6 +128,26 @@ public AcLoadFlowResult run() {
         voltageInitializer.prepare(context.getNetwork());
 
         RunningContext runningContext = new RunningContext();
+        double distributedActivePower = 0.0;
+        ReportNode reportNode = context.getNetwork().getReportNode();
+
+        // Verify whether a regulated bus voltage exists.
+        // If not, then fail immediately with SOLVER_FAILED status.
+        // Note that this approach is not perfect and could be improved in the future, in particular in
+        // security analysis and remedial actions context.
+        // For example: a contingency may cause the last PV node to disappear. In this case here we would
+        // just report SOLVER_FAILED. However, there could be other generators blocked at MinQ or MaxQ that
+        // _could potentially_ recover the situation, but this will not be tried at all...
+        boolean hasVoltageRegulatedBus = context.getEquationSystem().getEquations()
```

Comment:
I need to understand the cost of this iteration and if it is acceptable since we go through this at each run of each contingency and each conex component.

Reply:
The duration of this test is negligible from my measures (<1ms) even if sometimes you need to iterate several thousands of time to find the first active PV bus.

Reply:
But this seems much faster
boolean hasVoltageRegulatedBus = context.getNetwork().getBuses().stream().anyMatch(b -> b.isGeneratorVoltageControlEnabled() && !b.isDisabled() && !b.getGeneratorVoltageControl().get().isDisabled())

Reply:
Thank you, I changed with your better way